clippy_lints/src/recursive_format_impl.rs

```diff
@@ -0,0 +1,269 @@
+use clippy_utils::diagnostics::span_lint;
+use clippy_utils::macros::{FormatArgsArg, FormatArgsExpn};
+use clippy_utils::{is_diag_trait_item, match_def_path, path_to_local_id, paths};
+use if_chain::if_chain;
+use rustc_hir::{Expr, ExprKind, HirId, Impl, ImplItem, ImplItemKind, Item, ItemKind, UnOp};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_tool_lint, impl_lint_pass};
+use rustc_span::{sym, ExpnData, ExpnKind, Symbol};
+
+const FORMAT_MACRO_PATHS: &[&[&str]] = &[
+    &paths::FORMAT_ARGS_MACRO,
+    &paths::ASSERT_EQ_MACRO,
+    &paths::ASSERT_MACRO,
+    &paths::ASSERT_NE_MACRO,
+    &paths::EPRINT_MACRO,
+    &paths::EPRINTLN_MACRO,
+    &paths::PRINT_MACRO,
+    &paths::PRINTLN_MACRO,
+    &paths::WRITE_MACRO,
+    &paths::WRITELN_MACRO,
+];
+
+#[derive(Clone, Copy)]
+enum ImplTrait {
+    Debug,
+    Display,
+}
+
+const FORMAT_MACRO_DIAG_ITEMS: &[Symbol] = &[sym::format_macro, sym::std_panic_macro];
+
+fn outermost_expn_data(expn_data: ExpnData) -> ExpnData {
+    if expn_data.call_site.from_expansion() {
+        outermost_expn_data(expn_data.call_site.ctxt().outer_expn_data())
+    } else {
+        expn_data
+    }
+}
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for recursive use of `Display` or `Debug` traits inside their implementation.
+    ///
+    /// ### Why is this bad?
+    /// This is unconditional recursion and so will lead to infinite
+    /// recursion and a stack overflow.
+    ///
+    /// ### Example
+    ///
+    /// ```rust
+    /// use std::fmt;
+    ///
+    /// struct Structure(i32);
+    /// impl fmt::Display for Structure {
+    ///     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+    ///         write!(f, "{}", self.to_string())
+    ///     }
+    /// }
+    ///
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// use std::fmt;
+    ///
+    /// struct Structure(i32);
+    /// impl fmt::Display for Structure {
+    ///     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+    ///         write!(f, "{}", self.0)
+    ///     }
+    /// }
+    /// ```
+    #[clippy::version = "1.48.0"]
+    pub RECURSIVE_FORMAT_IMPL,
+    correctness,
+    "Format trait method called while implementing the same Format trait"
+}
+
+#[derive(Default)]
+pub struct RecursiveFormatImpl {
+    // Whether we are inside Display or Debug trait impl - None for neither
+    format_trait_impl: Option<ImplTrait>,
+    // hir_id of self parameter of method inside Display Impl - i.e. fmt(&self)
+    self_hir_id: Option<HirId>,
```

Comment:
I think you can avoid storing this and just check whether the variable name is `kw::SelfLower`.

Reply:
Cool, this works. I was wondering if there's ever a case where this is less powerful? But it seems all the tests pass either way :slightly_smiling_face: